--- a/networking/ether-wake.c
+++ b/networking/ether-wake.c
@@ -115,10 +115,8 @@ static void get_dest_addr(const char *ho
 	if (eap) {
 		*eaddr = *eap;
 		bb_debug_msg("The target station address is %s\n\n", ether_ntoa(eaddr));
-#if !defined(__UCLIBC__)
 	} else if (ether_hostton(hostid, eaddr) == 0) {
 		bb_debug_msg("Station address for hostname %s is %s\n\n", hostid, ether_ntoa(eaddr));
-#endif
 	} else
 		bb_show_usage();
 }
